File: artifacts/go-api/internal/domain/experiment.go
package domain

import (
        "encoding/json"
        "time"
)

// ExperimentStatus mirrors the frontend enum.
type ExperimentStatus = string

const (
        StatusExploring    ExperimentStatus = "探索中"
        StatusReproducible ExperimentStatus = "可复现"
        StatusFailed       ExperimentStatus = "失败"
        StatusVerified     ExperimentStatus = "已验证"
)

// ExperimentRecord is a single experiment run inside a SciNote.
//
// CurrentModules stores OntologyModule[] as an opaque JSON blob — the Go
// backend does not parse the individual module structure.  When a single
// module key is updated (PATCH /api/experiments/:id/modules/:key), the
// service layer merges the new module into the existing JSON array in the
// database without full re-serialization of all modules.
//
// Confirmation-state and lineage fields are described in domain/inheritance.go.
type ExperimentRecord struct {
        ID                 string
        SciNoteID          string
        Title              string
        PurposeInput       *string
        ExperimentStatus   ExperimentStatus
        ExperimentCode     string
        Tags               []string
        EditorContent      string
        ReportHtml         *string
        // Report persistence metadata (added Phase 2 — migration 20260319006)
        ReportGeneratedAt  *time.Time
        ReportSource       *string         // "stub" | "ai" | "manual" | nil (no report)
        ReportUpdatedAt    *time.Time
        ReportModelJson    json.RawMessage // ExperimentReportModel JSON snapshot
        CurrentModules     json.RawMessage // OntologyModule[] opaque blob
        InheritedVersionID *string         // forward ref to ontology_versions (Phase 2)
        IsDeleted          bool
        CreatedAt          time.Time
        UpdatedAt          time.Time

        // ---------------------------------------------------------------------------
        // Inheritance-chain fields (added in migration 20260315004)
        // ---------------------------------------------------------------------------

        // SequenceNumber is the 1-based ordinal position of this record within its SciNote.
        SequenceNumber int

        // ConfirmationState is one of: "draft", "confirmed", "confirmed_dirty".
        // Transitions are managed exclusively by ExperimentService.
        ConfirmationState ConfirmationState

        // ConfirmedAt is the timestamp of the most recent confirm action (nil until first confirm).
        ConfirmedAt *time.Time

        // ConfirmedModules is the heritable-modules snapshot stored at the time of
        // the last confirm.  Nil until first confirm.
        ConfirmedModules json.RawMessage

        // DerivedFromSourceType indicates whether this record's defaults came from
        // the SciNote's initial_modules ("initial") or from a prior confirmed record ("record").
        DerivedFromSourceType DerivedFromSourceType

        // DerivedFromRecordID is the ID of the record that provided the defaults.
        // Nil when DerivedFromSourceType == "initial".
        DerivedFromRecordID *string

        // DerivedFromRecordSeq is the sequence_number of the source record.
        // Nil when DerivedFromSourceType == "initial".
        // Stored redundantly so the frontend never needs a follow-up query to display banner text.
        DerivedFromRecordSeq *int

        // DerivedFromContextVer is the context_version of the parent SciNote at the
        // moment this record was created.  Used for audit / future diff features.
        DerivedFromContextVer int
}

// RecentExperimentRow is the result type for the cross-table recent-experiments query.
// It joins experiment_records with sci_notes to produce the fields needed by
// the home page feed without a frontend N+1 loop.
type RecentExperimentRow struct {
        ExperimentID     string
        ExperimentTitle  string
        SciNoteID        string
        SciNoteTitle     string
        ExperimentStatus ExperimentStatus
        CreatedAt        time.Time
        UpdatedAt        time.Time
}

// ExperimentPatch carries optional update fields for PATCH /api/experiments/:id.
// Only non-nil fields are written to the database.
// Tags nil = no change; Tags []string{} = clear to empty array.
// CurrentModules nil = no change; non-nil replaces the entire current_modules column.
type ExperimentPatch struct {
        Title            *string
        ExperimentStatus *string
        ExperimentCode   *string
        Tags             []string
        EditorContent    *string
        ReportHtml       *string
        CurrentModules   json.RawMessage // nil = no change; non-nil replaces whole column
}

// CreateExperimentInput carries the fields supplied by the caller for a new record.
// The service layer resolves inheritance and fills lineage fields before persisting.
type CreateExperimentInput struct {
        Title              string
        PurposeInput       *string
        ExperimentStatus   ExperimentStatus
        ExperimentCode     string
        Tags               []string
        EditorContent      string
        CurrentModules     json.RawMessage // frontend-supplied module array (used to bootstrap initial_modules)
        InheritedVersionID *string
}

File: artifacts/web/src/types/experiment.ts
/**
 * experiment.ts — domain types for ExperimentRecord and its lifecycle.
 *
 * Layer: types (pure data contracts, no runtime logic)
 *
 * Scope:
 *   - ExperimentRecord entity + all its lifecycle / chain fields
 *   - ConfirmationState, DerivedFromSourceType, ExperimentStatus enums
 *   - Wizard / new-experiment-flow types (WizardStep, StepAiStatus, etc.)
 *
 * Ontology UI model types (OntologyModule, OntologyVersion, etc.) live in
 * types/workbench.ts because they belong to the workbench UI layer.
 */

// ---------------------------------------------------------------------------
// Wizard / new-experiment-flow types
// ---------------------------------------------------------------------------

export interface WizardStep {
  id: number;
  label: string;
}

/**
 * AI analysis / generation status for an individual wizard step.
 *
 * "idle"       — not yet touched by AI (default)
 * "processing" — AI is currently analyzing / generating content for this step
 * "generated"  — content has been auto-generated; user has not yet reviewed it
 * "reviewed"   — user navigated to this step after AI generation
 *
 * State machine (one-way):
 *   idle → processing → generated → reviewed
 */
export type StepAiStatus = "idle" | "processing" | "generated" | "reviewed";

/** Immutable map from stepId → StepAiStatus */
export type StepAiStatusMap = ReadonlyMap<number, StepAiStatus>;

export type FileStatus = "pending" | "analyzing" | "done";

export interface ImportedFile {
  id: string;
  name: string;
  fileType: string;   // e.g. "PDF", "DOCX"
  size: string;       // human-readable
  importedAt: string; // display string, e.g. "14:32"
  status: FileStatus;
}

// ---------------------------------------------------------------------------
// ExperimentRecord — confirmation chain fields
// ---------------------------------------------------------------------------

/**
 * Three-state lifecycle of an ExperimentRecord.
 * Managed exclusively by the server; the frontend treats this as read-only.
 *
 *   draft           — created, never confirm-saved.
 *   confirmed       — confirm-saved at least once; heritable modules are locked in.
 *   confirmed_dirty — confirmed previously, but modules edited since last confirm.
 */
export type ConfirmationState = "draft" | "confirmed" | "confirmed_dirty";

/**
 * Whether a record's heritable defaults came from the SciNote's immutable
 * initial_modules ("initial") or from a prior confirmed record ("record").
 */
export type DerivedFromSourceType = "initial" | "record";

// ---------------------------------------------------------------------------
// ExperimentRecord entity
// ---------------------------------------------------------------------------

export type ExperimentStatus = "探索中" | "可复现" | "失败" | "已验证";

export const EXPERIMENT_STATUS_OPTIONS: ExperimentStatus[] = [
  "探索中",
  "可复现",
  "失败",
  "已验证",
];

export interface ExperimentRecord {
  id: string;
  sciNoteId: string;
  title: string;
  purposeInput?: string;
  experimentStatus: ExperimentStatus;
  experimentCode: string;
  tags: string[];
  /** Which ontology version this record inherited from (legacy / forward-ref field) */
  inheritedOntologyVersionId: string;
  /** Live working copy of modules — may diverge from the inherited version */
  currentModules: import("./workbench").OntologyModule[];
  /** TipTap HTML content */
  editorContent: string;
  createdAt: string;
  /** Server-assigned last-modification timestamp. Used e.g. for deletedAt in trash. */
  updatedAt?: string;
  /**
   * AI-generated experiment report HTML.
   * Phase 1: locally generated (stub).
   * Phase 2+: server-persisted via POST /api/experiments/:id/report/generate.
   * Undefined until all five modules are confirmed and the report is generated.
   */
  reportHtml?: string;

  /**
   * Who/what produced the current reportHtml.
   * "stub"   → Phase-1 local rule-based generation (legacy, may exist in DB)
   * "ai"     → Phase-2 server-side AI generation
   * "manual" → user edited and explicitly saved the report
   * undefined → no report exists yet
   */
  reportSource?: string;

  /**
   * ISO timestamp when the report was first generated (generate endpoint sets this).
   * Undefined until first generation.
   */
  reportGeneratedAt?: string;

  /**
   * ISO timestamp of the last save (generation or manual edit).
   * Undefined until first report save.
   */
  reportUpdatedAt?: string;

  // ---------------------------------------------------------------------------
  // Inheritance-chain fields (server-assigned; read-only in frontend)
  // ---------------------------------------------------------------------------

  /** 1-based ordinal position of this record within its SciNote. */
  sequenceNumber: number;

  /** Confirmation lifecycle state. Managed by the server. */
  confirmationState: ConfirmationState;

  /** ISO timestamp of the most recent confirm-save action. Undefined until first confirm. */
  confirmedAt?: string;

  /**
   * Where this record's heritable defaults came from.
   * "initial"  → seeded from the SciNote's wizard initialization modules.
   * "record"   → inherited from the last confirmed record in the chain.
   */
  derivedFromSourceType: DerivedFromSourceType;

  /**
   * The sequence_number of the record this record inherited from.
   * Undefined when derivedFromSourceType === "initial".
   * Used for the banner: "已继承第N条确认保存的记录"
   */
  derivedFromRecordSeq?: number;

  /**
   * The ID of the record this record inherited from.
   * Undefined when derivedFromSourceType === "initial".
   */
  derivedFromRecordId?: string;

  /** context_version of the parent SciNote at creation time (audit field). */
  derivedFromContextVer: number;
}

// ---------------------------------------------------------------------------
// AI title assist
// ---------------------------------------------------------------------------

export interface PurposeAssistResult {
  generatedTitle: string;
  purposeDraft: string;
  highlightedModuleKeys: import("./workbench").OntologyModuleKey[];
}
